Desktop/results/ploteditormodel.cpp

```diff
@@ -134,6 +143,86 @@ void PlotEditorModel::somethingChanged()
 	}
 }
 
+void PlotEditorModel::undoSomething()
+{
+	if (!_undo.empty())
+	{
+		_redo.push(_imgOptions);
+
+		_imgOptions = _undo.top();
+		_undo.pop();
+
+		setRedoEnabled(true);
+		setUndoEnabled(!_undo.empty());
+
+		applyChangesFromUndoOrRedo();
+	}
+}
+
+void PlotEditorModel::redoSomething()
+{
+	if (!_redo.empty())
+	{
+		_undo.push(_imgOptions);
+
+		_imgOptions = _redo.top();
+		_redo.pop();
+
+		setRedoEnabled(!_redo.empty());
+		setUndoEnabled(true);
+
+		applyChangesFromUndoOrRedo();
+	}
+}
+
+void PlotEditorModel::applyChangesFromUndoOrRedo()
+{
+	setLoading(true);
+	_editOptions = _imgOptions["editOptions"];
+
+	_xAxis->setAxisData(_editOptions["xAxis"]);
+	_yAxis->setAxisData(_editOptions["yAxis"]);
+
+	_prevImgOptions = _imgOptions;
+	_analysis->editImage(_prevImgOptions);
+	setLoading(false);
+}
+
+void PlotEditorModel::setUndoEnabled(bool undoEnabled)
+{
+	if (_undoEnabled == undoEnabled)
+		return;
+
+	_undoEnabled = undoEnabled;
+	emit undoEnabledChanged(_undoEnabled);
+}
+
+void PlotEditorModel::setRedoEnabled(bool redoEnabled)
```

Comment:
This whole function isn't necessary and I'll show you later how it can be done more efficiently.

Reply:
(And avoiding double administration, because redo-enabled == redostacksize > 0 )